### PR TITLE
Client server SSL host verification

### PR DESCRIPTION
**Pull request by manics**
See https://trello.com/c/J5DFWEzE/593-how-to-prevent-mitm-for-users

--depends-on https://github.com/openmicroscopy/openmicroscopy/pull/5998

### CODE REVIEW

File: omero/sysadmins/client-server-ssl.rst [hunk 1]
```diff
@@ -0,0 +1,86 @@
+Client Server SSL verification
+==============================
+
+If you configure OMERO.web behind Nginx with a recognised SSL certificate your users can be sure that they are connecting to their intended server.
```
Comment (mtbc) [inline omero/sysadmins/client-server-ssl.rst]: recognized

File: omero/sysadmins/client-server-ssl.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+This may result in users inadvertently transmitting their login credentials to an attacker.
+
+This can be remedied by configuring OMERO.server with a certificate (the same certificate used for OMERO.web Nginx may work), and ensuring all OMERO clients are configured to verify the server certificate before connecting.
+
+
+Server certificate
+------------------
+
+The easiest solution is to re-use the SSL certificates used to protect OMERO.web.
+First convert the public certificate ``server.pem`` and private key ``server.key`` to the PKCS12 format where ``secret`` is the password used to protect the combined output file ``server.p12``::
```
Comment (mtbc) [inline omero/sysadmins/client-server-ssl.rst]: We have `:file:` markup available.

File: omero/sysadmins/client-server-ssl.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    # Look for certificates in this directory, you can omit and use the full path to files instead
+    omero config set omero.glacier2.IceSSL.DefaultDir /etc/ssl/omero/
+    omero config set omero.glacier2.IceSSL.CertFile server.p12
+    omero config set omero.glacier2.IceSSL.Password secret
+
+For even stronger security require TLS 1.2, disable anonymous ciphers and only allow ``HIGH``::
+
+    omero config set omero.glacier2.IceSSL.Protocols tls1_2
+    omero config set omero.glacier2.IceSSL.ProtocolVersionMin tls1_2
+    omero config set omero.glacier2.IceSSL.ProtocolVersionMax tls1_2
```
Comment (mtbc) [inline omero/sysadmins/client-server-ssl.rst:37]: Is setting a maximum beneficial?

Comment (manics) [reply omero/sysadmins/client-server-ssl.rst:37]: It's currently set to `tls1_1`: https://github.com/openmicroscopy/openmicroscopy/blob/v5.5.0-m5/etc/templates/grid/templates.xml#L485 and I'm not sure what happens if `Protocols` or `ProtocolVersionMin` is greater than `ProtocolVersionMax`

Comment (mtbc) [reply omero/sysadmins/client-server-ssl.rst:37]: Ahh, it must be set to *something*? Makes perfect sense, ta.

Comment (manics) [reply omero/sysadmins/client-server-ssl.rst:37]: More that it's set by default in openmicroscopy, and I can't think of an easy way to unset it.

**Comment by mtbc** [comment]
Staged at https://ci.openmicroscopy.org/job/OMERO-DEV-merge-docs/ws/src/omero/_build/html/sysadmins/client-server-ssl.html.

**Comment by manics** [comment]
Well that's unexpected, I thought as a draft PR it wouldn't be!

**Comment by mtbc** [comment]
I too was surprised but at least it's convenient in this case.

File: omero/sysadmins/client-server-ssl.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+`Zeroc provide the Ice Certificate Utilities package <https://pypi.org/project/zeroc-icecertutils/>`_ to help create certificates, but if you know what you are doing you can use ``openssl`` directly.
+
+
+Client host verification
+------------------------
+
+At present there is no easy way to configure the standard OMERO clients to require host verification.
+
+If you are a developer the following Ice properties can be passed to the ``omero.client`` constructor to force host validation:
```
Comment (mtbc) [inline omero/sysadmins/client-server-ssl.rst:60]: Is there a way to do this via `BlitzGateway`?

**Comment by mtbc** [comment]
Gah, I'm running into "IceSSL: unable to establish CA certificates". `iceca init` made me a `ca.pem` and whatnot so I naively did a `iceca export --alias ca ~/omero/ca.crt` then passed `--IceSSL.CAs=/home/mtbc/omero/ca.crt` but maybe that's complete nonsense?

File: omero/sysadmins/client-server-ssl.rst (hunk 4 above)
Comment (manics) [reply omero/sysadmins/client-server-ssl.rst:60]: I don't know if you can pass them directly to the BlitzGateway. You can construct the BlitzGateway from an omero.client object though:

```
iceargs = {
    # 'Ice.Trace.Network': '3',
    # 'Ice.Trace.Protocol': '1',
    # 'IceSSL.Trace.Security': '1',
    'IceSSL.VerifyPeer': '1',
    'IceSSL.VerifyDepthMax': 0,
    # 'IceSSL.UsePlatformCAs': '1',
    # 'IceSSL.CheckCertName': '1',
    'IceSSL.Protocols': 'tls1_2',
    'IceSSL.Ciphers': 'HIGH',
}
client = omero.client(host=host, args=['--%s=%s' % a for a in iceargs.items()])
conn = BlitzGateway(client_obj=client)
```

**Comment by mtbc** [comment]
(Of course, the `CertFile` I point the server to was created with `iceca create`.)

**Comment by mtbc** [comment]
On Debian 9.8 with,

```
iceca init
iceca create localhost
iceca export --alias localhost --password omero ~/omero/server.p12
```

```
Ice.Default.Host=localhost
omero.glacier2.IceSSL.CertFile=server.p12
omero.glacier2.IceSSL.Ciphers=HIGH
omero.glacier2.IceSSL.DefaultDir=/home/mtbc/omero/
omero.glacier2.IceSSL.Password=omero
omero.glacier2.IceSSL.ProtocolVersionMax=tls1_2
omero.glacier2.IceSSL.ProtocolVersionMin=tls1_2
omero.glacier2.IceSSL.Protocols=tls1_2
```

```python
import omero

client = omero.client(['--IceSSL.Ciphers=HIGH',
                       '--IceSSL.VerifyPeer=1',
                       '--IceSSL.VerifyDepthMax=0',
                       '--IceSSL.CAs=/home/mtbc/.iceca/ca.pem',
                       '--IceSSL.Protocols=tls1_2',
                       '--IceSSL.CheckCertName=1',
                       'localhost'])
try:
    factory = client.createSession('user-2', 'ome')
    query_service = factory.getQueryService()
    query = 'SELECT name FROM Dataset'
    for path in query_service.projection(query, None):
        print path[0].val
finally:
    client.closeSession()
```

I can connect just fine! :fireworks:

File: omero/sysadmins/client-server-ssl.rst (hunk 4 above)
Comment (mtbc) [reply omero/sysadmins/client-server-ssl.rst:60]: That's great, very handy to know about, ta.

**Comment by joshmoore** [comment]
@manics : is this scenario covered by https://github.com/ome/docker-example-omero-websockets/pull/1? Or do you see having a more focused example?

**Comment by manics** [comment]
No. It could easily be added to the same docker-compose though if you think it's not going to confuse people?

**Comment by joshmoore** [comment]
Happy to have it as a second repo (e.g. `docker-example-omero-servercert`) then.

**Comment by manics** [comment]
Rebased without changes, moved to additional/advanced server features.

**Comment by joshmoore** [comment]
No objections today during standup. Merging. Thanks, @manics

**Pull request merged**